packer/plugin.go

```diff
@@ -268,4 +269,26 @@ type PluginDetails struct {
 	PluginPath  string
 }
 
-var PluginsDetailsStorage = map[string]PluginDetails{}
+type PluginsDetailsStorage struct {
+	rwMutex sync.RWMutex
+	data    map[string]PluginDetails
+}
+
+var GlobalPluginsDetailsStore = &PluginsDetailsStorage{
+	data: make(map[string]PluginDetails),
+}
+
+func (pds *PluginsDetailsStorage) Set(key string, plugin PluginDetails) {
+	pds.rwMutex.Lock()
+	defer pds.rwMutex.Unlock()
+	if _, exists := pds.data[key]; !exists {
+		pds.data[key] = plugin
+	}
+}
+
+func (pds *PluginsDetailsStorage) Get(key string) (PluginDetails, bool) {
+	pds.rwMutex.RLock()
+	defer pds.rwMutex.RUnlock()
+	plugin, exists := pds.data[key]
+	return plugin, exists
+}
```

Comment:
So on the Set and Get methods one of the concerns was having to construct the key using component type constant and name. Which I agree can be error prone.

As an alternative what about having setters and getters for the three component types. 

```suggestion
func (pds *PluginsDetailsStorage) SetBuilder(name string, value PluginDetails) {
       key := fmt.Sprintf("%q-%q", PluginComponentBuilder, name)
       return pds.set(key, value)
}

func (pds *PluginsDetailsStorage) set(key string, plugin PluginDetails) {
	pds.rwMutex.Lock()
	defer pds.rwMutex.Unlock()
	if _, exists := pds.data[key]; !exists {
		pds.data[key] = plugin
	}
}

func (pds *PluginsDetailsStorage) GetBuilder(name string) (PluginDetails, bool) {
       key := fmt.Sprintf("%q-%q", PluginComponentBuilder, name)
       return pds.get(key)
}

func (pds *PluginsDetailsStorage) get(key string) (PluginDetails, bool) {
	pds.rwMutex.RLock()
	defer pds.rwMutex.RUnlock()
	plugin, exists := pds.data[key]
	return plugin, exists
}
```
Not only does it consolidate the logic for building the key into a single location. It also makes it easier to change implementation if needed.